File: figures/06-search-trees/snippet-avl-tree.java
    private static final class Node {int key,height;Node left,right;Node(int k){key=k;}}
    private Node root;private int size;
    private static int h(Node x){return x==null?-1:x.height;}
    private static void update(Node x){x.height=1+Math.max(h(x.left),h(x.right));}
    private static int bf(Node x){return h(x.left)-h(x.right);}
    public boolean contains(int k){Node x=root;while(x!=null){if(k==x.key)return true;x=k<x.key?x.left:x.right;}return false;}
    public boolean insert(int k){if(contains(k))return false;root=insert(root,k);size++;return true;}
    private static Node insert(Node x,int k){if(x==null)return new Node(k);if(k<x.key)x.left=insert(x.left,k);else x.right=insert(x.right,k);return rebalance(x);}
    private static Node rebalance(Node x){update(x);if(bf(x)>1){if(bf(x.left)<0)x.left=left(x.left);return right(x);}if(bf(x)<-1){if(bf(x.right)>0)x.right=right(x.right);return left(x);}return x;}
    private static Node right(Node y){Node x=y.left,b=x.right;x.right=y;y.left=b;update(y);update(x);return x;}
    private static Node left(Node x){Node y=x.right,b=y.left;y.left=x;x.right=b;update(x);update(y);return y;}
    public List<Integer> inorder(){List<Integer>o=new ArrayList<>();inorder(root,o);return List.copyOf(o);}
    private static void inorder(Node x,List<Integer>o){if(x!=null){inorder(x.left,o);o.add(x.key);inorder(x.right,o);}}
    public int height(){return h(root);}

File: figures/06-search-trees/snippet-avl-tree.py
class _Node:
    def __init__(self, key):
        self.key = key
        self.height = 0
        self.left = self.right = None


def _h(x):
    return -1 if x is None else x.height


def _update(x):
    x.height = 1 + max(_h(x.left), _h(x.right))


def _bf(x):
    return _h(x.left) - _h(x.right)


def _rotate_right(y):
    x = y.left
    b = x.right
    x.right = y
    y.left = b
    _update(y)
    _update(x)
    return x


def _rotate_left(x):
    y = x.right
    b = y.left
    y.left = x
    x.right = b
    _update(x)
    _update(y)
    return y


def _rebalance(x):
    _update(x)
    if _bf(x) > 1:
        if _bf(x.left) < 0:
            x.left = _rotate_left(x.left)
        return _rotate_right(x)
    if _bf(x) < -1:
        if _bf(x.right) > 0:
            x.right = _rotate_right(x.right)
        return _rotate_left(x)
    return x


class AVLTree:
    def __init__(self):
        self.root = None
        self._size = 0

    def __len__(self):
        return self._size

    def contains(self, key):
        x = self.root
        while x is not None:
            if key == x.key:
                return True
            x = x.left if key < x.key else x.right
        return False

    def insert(self, key):
        if self.contains(key):
            return False
        self.root = self._insert(self.root, key)
        self._size += 1
        return True

    def _insert(self, x, key):
        if x is None:
            return _Node(key)
        if key < x.key:
            x.left = self._insert(x.left, key)
        else:
            x.right = self._insert(x.right, key)
        return _rebalance(x)

    def height(self):
        return _h(self.root)

    def inorder(self):
        out = []

        def walk(x):
            if x is not None:
                walk(x.left)
                out.append(x.key)
                walk(x.right)

        walk(self.root)
        return out
